src/JSInterop/Microsoft.JSInterop/src/DotNetObjectRefOfT.cs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.JSInterop
+{
+    /// <summary>
+    /// Wraps a JS interop argument, indicating that the value should not be serialized as JSON
+    /// but instead should be passed as a reference.
+    ///
+    /// To avoid leaking memory, the reference must later be disposed by JS code or by .NET code.
+    /// </summary>
+    /// <typeparam name="TValue">The type of the value to wrap.</typeparam>
+    public sealed class DotNetObjectRef<TValue> : IDotNetObjectRef, IDisposable where TValue : class
+    {
+        private long? _trackingId;
+
+        /// <summary>
+        /// This A is for meant for JSON deserialization and should not be used by user code.
+        /// </summary>
+        [Obsolete("This API is meant for JSInterop infrastructure and should not be used by user code.")]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public DotNetObjectRef()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="DotNetObjectRef{TValue}" />.
+        /// </summary>
+        /// <param name="value">The value to pass by reference.</param>
+        public DotNetObjectRef(TValue value)
+        {
+            Value = value;
+            _trackingId = DotNetObjectRefManager.Current.TrackObject(this);
+        }
+
+        /// <summary>
+        /// Gets the object instance represented by this wrapper.
+        /// </summary>
+        [JsonIgnore]
+        public TValue Value
+        {
+            get;
+            // Workaround for https://github.com/dotnet/corefx/issues/37536 and https://github.com/dotnet/corefx/issues/37567
+            // Once fixed, we should make this property private.
+            set;
+        }
+
+        internal bool Disposed { get; private set; }
+
+        /// <summary>
+        /// This constructor is for meant for JSON serialization and should not be used by user code.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [JsonPropertyName(DotNetDispatcher.DotNetObjectRefKey)]
+        [Obsolete("This API is meant for JSInterop infrastructure and should not be used by user code.")]
+        public long TrackingId
```

Comment:
Is there a longer-term plan for the JSON serializer to support serializing nonpublic fields given suitable metadata?

If not, we might consider actually calling the property `__dotNetObject` to make it extra unappealing.

Reply:
> Is there a longer-term plan for the JSON serializer to support serializing nonpublic fields given suitable metadata?

Not entirely sure, but we I'll all for making it unappealing 👍

Reply:
If we need that (and I think getting rid of a weird `Obsolete` public property is worth it) then we should file it and raise it with them.

Reply:
@anurse it's likely that we'll use converters for this once the feature lands. As of now, having an ugly property that isn't visible to user code isn't super awful.

Reply:
Cool, as long as we have *a* strategy for getting on the right path again and any dependencies that strategy has are tracked I'm happy :)